File: java/2025/3/partition_labels/PartitionLabels.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.stream.Collectors;

public class PartitionLabels {

    /* Buscamos nesse problema separar a string no maior numero de partições onde as letras de uma partição
        não se existam em nenhuma outra partição.
       Basta criar a partição da letra corrente e verificar se há algum letra de dentro dessa partição em um local
        mais distante na string, se sim, aumentar a partição até a ultima ocorrecia dessas ocorrencias.

        n = s.length

        Time: O(n^2)
        Space (n)
    */

    public static List<Integer> partitionLabels(String s) {
        List<String> partitions = new ArrayList<>();
        int n = s.length();
        int i = 0;

        while (i < n) { //O(n)
            //Busca ultimo index da letra
            char current = s.charAt(i);
            int lastIndex = s.lastIndexOf(current);

            //Busca ultimo index real do trecho selecionado
            for (int j = i + 1; j < lastIndex; j++) { //* O(n)
                char inner = s.charAt(j);
                int innerLastIndex = s.lastIndexOf(inner);
                if(innerLastIndex > lastIndex) {
                    lastIndex = innerLastIndex;
                }
            }

            //Adiciona a particao encontrada
            partitions.add(s.substring(i, lastIndex+1));
            i = lastIndex+1;
        }

        return partitions.stream()
                .map(String::length)
                .collect(Collectors.toList());
    }

    // ===== TEST CASES =====
    public static void main(String[] args) {
        System.out.println("=== Partition Labels Tests ===");

        testPartition("ababcbacadefegdehijhklij", Arrays.asList(9,7,8), 1);

        testPartition("eccbbbbdec", Arrays.asList(10), 2);

        testPartition("abcdefg", Arrays.asList(1,1,1,1,1,1,1), 3);

        testPartition("aabbaabb", Arrays.asList(8), 4);

        testPartition("", Arrays.asList(), 5);
    }

    private static void testPartition(String input, List<Integer> expected, int testId) {
        System.out.printf("\nTest Case %d:%nInput: \"%s\"%n", testId, input);

        long startTime = System.nanoTime();
        List<Integer> result = partitionLabels(input);
        long endTime = System.nanoTime();

        System.out.println("Expected: " + expected);
        System.out.println("Result:   " + result);
        System.out.printf("Time: %d μs%n", (endTime - startTime) / 1000);

        if (result.equals(expected)) {
            System.out.println("✅ PASS");
        } else {
            System.out.println("❌ FAIL");
        }
    }
}

File: python/2025/3/partition_labels/partition_labels.py
import time
from typing import List

'''
    Reescrita em python do problema desenvolvido em java.
'''


def main() -> None:
    print("=== Partition Labels Tests ===")
    test_cases = [
        ("ababcbacadefegdehijhklij", [9, 7, 8], 1),
        ("eccbbbbdec", [10], 2),
        ("abcdefg", [1, 1, 1, 1, 1, 1, 1], 3),
        ("aabbaabb", [8], 4),
        ("", [], 5)
    ]

    for s, expected, test_id in test_cases:
        testPartition(s, expected, test_id)


def testPartition(input_str: str, expected: List[int], testId: int) -> None:
    print(f"\nTest Case {testId}:\nInput: {input_str}")

    start_time = time.perf_counter_ns()
    result = partitionLabels(input_str)
    elapsed_time = time.perf_counter_ns() - start_time

    print(f"Expected: {expected}")
    print(f"Result:   {result}")
    print(f"Time: {elapsed_time / 1000:.3f} μs")
    print("✅ PASS" if result == expected else "❌ FAIL")


def partitionLabels(s: str) -> List[int]:
    partitions = []
    n = len(s)
    i = 0

    while i < n:
        current = s[i]
        last_index = s.rfind(current)

        for j in range(i + 1, last_index):
            last_index = max(last_index, s.rfind(s[j]))

        partitions.append(s[i:last_index + 1])
        i = (last_index + 1)

    return [len(partition) for partition in partitions]


if __name__ == "__main__":
    main()
